src/coverlet.core/Helpers/InstrumentationHelper.cs

```diff
@@ -189,25 +182,57 @@ public bool PortablePdbHasLocalSource(string module, out string firstNotFoundDoc
-                Guid languageGuid = metadataReader.GetGuid(document.Language);
-
-                // We verify all docs and return false if not all are present in local
-                // We could have false negative if doc is not a source
-                // Btw check for all possible extension could be weak approach
-                // We exlude from the check the autogenerated source file(i.e. source generators)
-                // We exclude special F# construct https://github.com/coverlet-coverage/coverlet/issues/1145
-                if (!_fileSystem.Exists(docName) && !IsGeneratedDocumentName(docName) &&
-                    !IsUnknownModuleInFSharpAssembly(languageGuid, docName))
+                bool anyDocumentMatches = MatchDocumentsWithSourcesMissingAll(metadataReader);
+                if (!anyDocumentMatches)
                 {
-                    return (false, docName);
+                    _logger.LogVerbose($"Unable to instrument module: {module}, pdb without any local source files");
+                    return false;
                 }
             }
+
+            if (excludeAssembliesWithoutSources.Equals(AssemblySearchType.MissingAny))
+            {
+                (bool allDocumentsMatch, string notFoundDocument) = MatchDocumentsWithSourcesMissingAny(metadataReader);
+
+                if (!allDocumentsMatch)
+                {
+                    _logger.LogVerbose(
+                        $"Unable to instrument module: {module}, pdb without local source files, [{FileSystem.EscapeFileName(notFoundDocument)}]");
+                    return false;
+                }
+            }
+
+            return true;
+        }
+
+        private IEnumerable<(string documentName, bool documentExists)> DocumentSourceMap(MetadataReader metadataReader)
+        {
+            return metadataReader.Documents.ToList().Select(docHandle =>
+            {
+                Document document = metadataReader.GetDocument(docHandle);
+                string docName = _sourceRootTranslator.ResolveFilePath(metadataReader.GetString(document.Name));
+                return (docName, _fileSystem.Exists(docName));
+            });
+        }
+
+        private bool MatchDocumentsWithSourcesMissingAll(MetadataReader metadataReader)
+        {
+            var documentSourceMap = DocumentSourceMap(metadataReader).ToList();
+            return documentSourceMap.Any(x => x.documentExists.Equals(true));
+        }
+
+        private (bool allDocumentsMatch, string notFoundDocument) MatchDocumentsWithSourcesMissingAny(
+            MetadataReader metadataReader)
+        {
+            var documentSourceMap = DocumentSourceMap(metadataReader).ToList();
+
+            if (documentSourceMap.Any(x => x.documentExists.Equals(false)))
+                return (false, documentSourceMap.FirstOrDefault(x => x.documentExists.Equals(false)).documentName);
```

Comment:
Same as above, can work on the enumerable:

```suggestion
            var missingDoc = DocumentSourceMap(metadataReader).FirstOrDefault(x => !x.documentExists);
            if (missingDoc != null)
            {
                return (false, missingDoc.documentName);
            }
```

Reply:
I'm not so sure about this one. `FirstOrDefault` isn't meant to check whether a value exists or not, although we all (ab)use it this way. In this case `FirstOrDefault` would return a `ValueTuple<string, bool>` which defaults to `(documentName: null, documentExists: false)`.
I now could adapt it to sth. like `if (missingDoc.documenName != null){...}` but now it somehow reads like: "If there is a document whose name can't be resolved lets include the assembly." 
Even tho' this is just because all documents have a matching source file.

Reply:
I missed the tuple data type here, so ignore this comment :)

Reply:
simplified the predicate